vllm/v1/worker/utils.py

```diff
@@ -277,3 +277,27 @@ def bind_kv_cache(
     for layer_name, kv_cache in kv_caches.items():
         # NOTE: Use list because of v0 PP virtual engine.
         forward_context[layer_name].kv_cache = [kv_cache]
+
+
+def is_residual_scattered(vllm_config: VllmConfig,
+                          num_input_tokens: int) -> bool:
+    """Check if the residual tensor is scattered for sequence parallelism.
+
+    The residual tensor is scattered across tensor parallel ranks when sequence
+    parallelism and tensor parallelism is enabled, and the number of
+    input tokens is one of the compilation sizes.
+    """
+    if not vllm_config.compilation_config.pass_config.\
+        enable_sequence_parallelism:
+        return False
+
+    tp = vllm_config.parallel_config.tensor_parallel_size
+    # When sequence parallelism is enabled, we always pad num_input_tokens
+    # to be a multiple of tensor_parallel_size (tp) earlier.
+    if tp > 1:
+        assert num_input_tokens % tp == 0
+
+    is_residual_scattered = (tp > 1 and num_input_tokens
+                             in vllm_config.compilation_config.compile_sizes)
+
+    return is_residual_scattered
```

Comment:
I think this would clarify the flow here:
```suggestion
    if tp == 1:
        return False

    assert num_input_tokens % tp == 0

    # Currently, SP is only enabled for static size fx graphs.
    return (num_input_tokens in vllm_config.compilation_config.compile_sizes)
```